Changes required for tickets creation functionality

ekoval opened the pull request:
Changes required for tickets creation functionality.

Commit:
Minor change in integration tests.

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Restored settings (probably required by Travis).

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Correct requirements.txt to satisfy Github CI.

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Bump version.

snitch commented:
:sparkles: No lint errors found. :sparkles:

RayeN commented on integration_tests/config.py.sample:
```diff
@@ -1,2 +1,3 @@
-TEAMSUPPORT_ORG_ID = 'test-org-id'
-TEAMSUPPORT_AUTH_KEY = 'test-auth-key'
+# Please get actual value of auth_key from configs and renamed to `config.py`.
```
renamed -> rename this file ?

RayeN commented on teamsupport/services.py:
```diff
@@ -40,18 +44,77 @@ def __init__(self, org_id, auth_token, **kwargs):
-    def get_tickets(self, query_params=None):
+    def search_tickets(self, query_params=None):
+        if query_params is None:
+            query_params = {}
```
one liner ->  `query_params = query_params or {}`

RayeN commented on teamsupport/services.py:
```diff
@@ -40,18 +44,77 @@ def __init__(self, org_id, auth_token, **kwargs):
+    def delete_contact(self, contact_id):
+        self.delete('Contacts/{}'.format(contact_id))
+
+    def create_ticket(self, first_name, email, category, name, description):
```
I think service methods should remain simple.  And we should glue needed functionality in `models.py`

Commit:
Move create ticket logic to Model level.

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Noop change to restart Travis build on GitHub.

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Disable pypy to make Travis work.

snitch commented:
:sparkles: No lint errors found. :sparkles:

RayeN commented on teamsupport/models.py:
```diff
@@ -23,6 +27,31 @@ def __init__(self, client, ticket_id=None, data=None):
         self.id = self.TicketID
 
+    @classmethod
+    def create(self, client, first_name, email, category, name, description):
```
self -> cls

RayeN commented on teamsupport/models.py:
```diff
@@ -23,6 +27,31 @@ def __init__(self, client, ticket_id=None, data=None):
+        ticket_xml = client.create_ticket(data)
+        ticket_id = ticket_xml.find('TicketID').text
+        client.set_ticket_description(ticket_id, description)
+        return Ticket(client, ticket_id)
```
Ticket -> cls

RayeN commented on teamsupport/models.py:
```diff
@@ -23,6 +27,31 @@ def __init__(self, client, ticket_id=None, data=None):
+        else:
+            # Otherwise - create new one.
+            contact = client.create_contact(FirstName=first_name, Email=email)
+        contact_id = contact.find('ContactID').text
```
Probably it make sense to create Contact model and move all this logic there.

RayeN commented on teamsupport/models.py:
```diff
@@ -23,6 +27,31 @@ def __init__(self, client, ticket_id=None, data=None):
+        }
+
+        ticket_xml = client.create_ticket(data)
+        ticket_id = ticket_xml.find('TicketID').text
```
I think it makes sense to initialize ticket with xml obtained from `create_ticket`

Commit:
Opmitize creation of Ticket from ticket_xml.

snitch commented:
:sparkles: No lint errors found. :sparkles:

RayeN commented on integration_tests/test_tickets.py:
```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
```
Think this part is redundant

RayeN commented on teamsupport/models.py:
```diff
@@ -23,6 +27,27 @@ def __init__(self, client, ticket_id=None, data=None):
+        contact = Contact.get(client, first_name, email)
+        if contact is None:
+            # Otherwise - create new one.
+            contact = Contact.create(client, first_name, email)
```
we could probably use `get_or_create`?

ekoval replied:
Personally I hate `get_or_create` name.

RayeN replied:
Maybe it sounds awkward, but at least it clearly explains what it does :)

stormf replied:
it exists because the pattern of get, check null, if so then create is quite common :)

RayeN commented on teamsupport/models.py:
```diff
@@ -23,6 +27,27 @@ def __init__(self, client, ticket_id=None, data=None):
         self.id = self.TicketID
 
+    @classmethod
+    def create(cls, client, first_name, email, category, name, description):
```
not related to your changes, but passing client everywhere looks unnecessary.  I would create  `get_teamsupport_service()` classmethod in  `XmlMolel` for that.

RayeN commented on teamsupport/models.py:
```diff
@@ -62,3 +87,28 @@ def __init__(self, client, user_id=None, data=None):
+            raise MissingArgumentError(
+                "__init__() needs either a 'contact_id' or 'data' argument "
+                '(neither given)')
+        self.id = self.ContactID
```
I don't think we need all that logic from Ticket model here.  I would say accepting `data` in constructor would be enough for us.  Besides it looks like we don't have `get_contact` method

RayeN commented on teamsupport/services.py:
```diff
@@ -40,16 +40,46 @@ def __init__(self, org_id, auth_token, **kwargs):
         return content
 
+    def search_contacts(self, **query_params):
+        response = self.get('Contacts/', params=query_params)
```
maybe it makes sense to use paths lower-cased everywhere, just for consistency?

RayeN commented on teamsupport/services.py:
```diff
@@ -40,16 +40,46 @@ def __init__(self, org_id, auth_token, **kwargs):
+            'tickets', root='Ticket', data=data, send_as_xml=True)
         return self.parse_xml_response(response)
 
+    def set_ticket_description(self, ticket_id, description):
```
Looks like a part of model

ekoval replied:
Answered below.

RayeN commented on teamsupport/services.py:
```diff
@@ -40,16 +40,46 @@ def __init__(self, org_id, auth_token, **kwargs):
+        self.update_ticket_action(
+            ticket_id, action_id, {'Description': description})
+
+    def get_ticket_description(self, ticket_id):
```
The same here - probably should be moved to model

ekoval replied:
I added this to a model, but left it in service as well. Setting description is so common task, and it is done so implicitly using raw TS calls that I believe it makes sense to have it in services as well as in models.

RayeN replied:
Still think it should go to models, but it's a personal opinion.  Even if it's a very common task, it's a kind of business logic which resides on a higher level then operations over ticket's actions. Somehow I got used to 1 API call 1 service action in our `services.py` (and all shortcuts, rules, etc in `models.py`).  Do we plan to use these actions outside of Ticket model?

ekoval replied:
I consider services and models as different levels of same API. If you need more flexibility and power, you can use services. If you want more readability and simplicity - you can use models. I disagree that service methods must always correspond API calls. Why? Also, this case is a quite specific one. It is clearly specified in TeamSupport API docs (totally incomplete btw) that the description is an action with certain ID. So, this is an API level, isn't it?

RayeN replied:
I'm not sure that if something is put in API docs it immediately becomes a part of API level.  I would rather treat it as "how you can achieve something via API", but not a part of API itself.  And I don't have a strong opinion against leaving this method on service level (besides it's only my own opinion :)).  I think other ppl might have other opinions about what  to put to each abstraction level.   But you didn't answer my question regarding the intention of usage of these shortcuts. Do we plan to use them directly on service level (outside of models)?

ekoval replied:
> Do we plan to use them directly on service level (outside of models)?

Who knows? This is a client, it can be used by different people in different ways.

blaix replied:
> Somehow I got used to 1 API call 1 service action in our services.py (and all shortcuts, rules, etc in models.py)

This is the pattern we should be following.

> I disagree that service methods must always correspond API calls. Why?

For the same reason we use any pattern: familiarity. Any developer working on or using our service clients know what to expect from service classes and models.

It also makes the tests more straightforward. Typically, we _only_ integration test service classes, because they are _only_ API calls. All of our business logic is _only_ unit tested, mocking the service class so we don't have to make real API calls to test our logic.

stefanor replied:
:+1: @blaix. The services.py is a python function for each backend service call. Higher order logic goes elsewhere.

ekoval replied:
Ok, will change this. However, I can't agree.

RayeN commented on teamsupport/services.py:
```diff
@@ -63,8 +93,9 @@ def update_ticket(self, ticket_id, data):
 
-    def get_ticket_actions(self, ticket_id):
-        response = self.get('Tickets/{0}/Actions'.format(ticket_id))
+    def get_ticket_actions(self, ticket_id, query_params=None):
```
In some places we accept query_params as `**query_params` and in others like `query_params=None`.  Probably it makes sense to be consistent?

ekoval replied:
Fixed.

Commit:
- Don't pass TeamSupport client to each call. User method (for class methods) and property for object methods instead.

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
URLs made lowercase;

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
- New way of library initialization;
- Move more logic to models level;
- Tests changes.

snitch commented:
:sparkles: No lint errors found. :sparkles:

blaix commented on CHANGELOG.rst:
```diff
@@ -1,6 +1,11 @@
 Changelog
 =========
 
+0.1.4 (2015-11-12)
```
Can we call this DEV and bump version in its own pull request?

stefanor replied:
Related: The changes look backwards-incompatible. But we (and presumably everyone else) aren't actually using this, yet. So, I'm not too worried about sane versioning.

ekoval replied:
> Can we call this DEV and bump version in its own pull request?

What do you mean, @blaix ? Should I remove all version-related changes from this PR, and file a new PR for them later?

blaix replied:
Yes. 

Explained elsewhere

ekoval replied:
Done.

blaix commented on CHANGELOG.rst:
```diff
@@ -1,6 +1,11 @@
+0.1.4 (2015-11-12)
+----------------
+
+-  Enhance Create Ticket functionality
```
can you be more specific?

stefanor commented on teamsupport/__init__.py:
```diff
@@ -1,11 +1,17 @@
+__all__ = (Action, TeamSupportService, Ticket, User, config)
+
+
+def init(org_id, auth_key):
```
`init` makes it sound like it's doing work. `configure`? `login`? (ok, that option also sounds like it's doing work...)

These could also be parameters to `TeamSupportService`, but then that would have to be passed to the other classes constructors.

ekoval replied:
As for me, `init` isn't worse than `configure`. It initializes library. Currently it just logs in, but this can easily change. Regarding `login` - I don't want  user to know that I log in to TeamSupport service inside - this looks like violation of encapsulation.

stefanor commented on teamsupport/models.py:
```diff
@@ -1,19 +1,31 @@
 
 
+TICKET_STATUS_NEW = '212203'
+TICKET_TYPE_SUPPORT = '35731'
```
wtf are these things? Where do they come from?

ekoval replied:
Constants come from TeamSupport API. When you create ticket, you can't specify status as "New". You have to specify ID of "New" taken from TicketStatuses table (resource).

stefanor replied:
I'm guessing they aren't values that would be useful to anyone else using this library. Which means they should move into configuration / consumer side of this,.

blaix replied:
assuming I understand this correctly: that these are IDs that are unique to our status values, in which case I agree with @stefanor

ekoval replied:
Configuration? I can't agree with this. I don't think this can change ever. They are just IDs from `TicketStatus` table and `TicketType` table.

stefanor replied:
Are you saying that anyone using this library will have the same values? If that's not true, it has to be configurable. Or determined by looking up statuses and types by name.

ekoval replied:
Same values (I hope) because they are primary keys from TicketStatuses/TicketType tables. I can't believe they can change. Regarding getting them dynamically - how to find them? Looking for "New" string? If's far less reliable than primary keys as for me.

stefanor replied:
> I can't believe they can change.

I'm suggesting that they are different for different TeamSupport customers. I.e. NEW for yola is a different ID than new for some other TeamSupport user.

> Regarding getting them dynamically - how to find them?

I'd assume a Status / Type enumeration API. And then matching (probably exactly) by title, yes.

I don't really see any other options, beyond ignoring this until someone else tries to use this library, and finds that it has hard-coded constants that are Yola-specific.

ekoval replied:
> I'd assume a Status / Type enumeration API. And then matching (probably exactly) by title, yes.

I see your point. I'll check this today. I just was sure these constants are global, now I'm not that sure :)

ekoval replied:
I don't see any other way to understand this except ask TeamSupport. Which I don't want to do, because we'll loose one more week waiting for answer.
P.S.
I'll send email with this question to TeamSupport support.

stefanor commented on teamsupport/models.py:
```diff
@@ -23,14 +35,41 @@ def __init__(self, client, ticket_id=None, data=None):
         self.id = self.TicketID
 
+    @classmethod
+    def create(cls, first_name, email, category, name, description):
```
That argument order is quite weird. Why is name so far from first_name? Is name the rest of the name? The entire name?

ekoval replied:
"name" means ticket name, not client name. Renamed to "title".

stefanor commented on teamsupport/models.py:
```diff
@@ -23,14 +35,41 @@ def __init__(self, client, ticket_id=None, data=None):
-            ) for action in actions.findall('Action')
-        ], wrap=False)
+        return QueryList([Action(data=action)
+                         for action in actions.findall('Action')], wrap=False)
```
I'd indent one more space. The `for` is inside the square brackets.

Commit:
More specific info in ChangeLog and ReadMe; intendation.

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Rename "name" parameter to "title".

snitch commented on teamsupport/models.py:
```diff
@@ -23,14 +35,42 @@ def __init__(self, client, ticket_id=None, data=None):
+            contact = Contact.create(first_name, email)
+
+        data = {
+            'Name': name,
```
F821 undefined name 'name'

Commit:
Fix undefined variable usage.

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Change create_contact signature;

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Contact.get() call fixed - it accepts only e-mail now. Searching by first name is too "fragile".

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Removed get/set_ticket_description from service level.

snitch commented:
:sparkles: No lint errors found. :sparkles:

RayeN commented on requirements.txt:
```diff
@@ -1,8 +1,9 @@
+mock<1.2.0
+funcsigs<0.5
+nose < 2.0.0
+pinocchio < 1.0.0
```
minor, but we don't put whitespaces around `<` in requirements

Commit:
Removed read credentails from config.py.

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Unify way we pass parameters to services methods.
Remove unused methods from services.

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Remove whitespaces from requirements.txt.

snitch commented:
:sparkles: No lint errors found. :sparkles:

RayeN commented on teamsupport/models.py:
```diff
@@ -23,14 +35,53 @@ def __init__(self, client, ticket_id=None, data=None):
+    def get_description(self):
+        ticket_actions = self.client.get_ticket_actions(
+            self.id, SystemActionTypeID=1)
+        if len(ticket_actions) >= 1:
```
`if ticket_actions:` as option

Commit:
Use constant instead of magic value.

snitch commented on teamsupport/models.py:
```diff
@@ -52,8 +103,8 @@ def __init__(self, client, ticket_id=None, action_id=None, data=None):
-    def __init__(self, client, user_id=None, data=None):
-        self.client = client
+    def __init__(self, user_id=None, data=None):
+        self.client = cls.get_client()
```
F821 undefined name 'cls'

Commit:
Remove User model at all. It is not used.

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Updated ChangeLog.

snitch commented:
:sparkles: No lint errors found. :sparkles:

ekoval commented:
:bell:

Commit:
Added ticket.contacts functionality.

ekoval commented:
@yola/service-engineering 
Guys. I love you all, but could we please finish reviewing this? It blocks two more tickets. Sorry for being so annoying.

stefanor commented:
Still waiting for https://github.com/yola/teamsupport-python/pull/10#discussion_r44970274 to be dealt with. They confirmed that these IDs aren't global.

ekoval commented:
> Still waiting for #10 (comment) to be dealt with. They confirmed that these IDs aren't global.

If they're global, hardcoded values seems fine. No?

> Eugene, 
> Those id values are global.
> 
> Victor

Commit:
- TicketStatuses and TicketTypes are retrieved dynamically, because they depend on ORG_ID;
- FormCategory removed from obligatory fields, because this is a custome field specific to Yola.

snitch commented:
:sparkles: No lint errors found. :sparkles:

blaix commented:
He said they are confirmed not global.

On Tuesday, November 24, 2015, Eugene Koval notifications@github.com
wrote:

> Still waiting for #10 https://github.com/yola/teamsupport-python/pull/10
> (comment) to be dealt with. They confirmed that these IDs aren't global.
> 
> If they're global, hardcoded values seems fine. No?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/yola/teamsupport-python/pull/10#issuecomment-159198207
> .

ekoval commented:
Yes, but the answer was "They're global".

ekoval commented:
I tried to register new account, and values are not global - they're specific to ORG_ID. Implementation updated.

stefanor commented:
> Yes, but the answer was "They're global".

And then I asked him for clarification and he said they were for Yola only. (He had misunderstood what we meant by Global)

ekoval commented:
Anyway, new version is ready for review.

Commit:
Remove version-related changes. Will post with another PR.

snitch commented:
:sparkles: No lint errors found. :sparkles:

RayeN commented on teamsupport/__init__.py:
```diff
@@ -1,11 +1,17 @@
 from __future__ import absolute_import
 
-from teamsupport.models import Action, Ticket, User
+import teamsupport.config as config
```
`from teamsupport import config` isn't it the same?

Commit:
Remove settings from config.
Use kwargs for ticket creation.

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Add one more test; optimized one import;

snitch commented:
:sparkles: No lint errors found. :sparkles:

RayeN commented on teamsupport/models.py:
```diff
@@ -23,23 +37,99 @@ def __init__(self, client, ticket_id=None, data=None):
+    @classmethod
+    def create(
+            cls, user_email, user_first_name, user_last_name,
+            title, description, **params):
```
I think we can fit 2 lines instead of 3

RayeN commented on teamsupport/models.py:
```diff
@@ -23,23 +37,99 @@ def __init__(self, client, ticket_id=None, data=None):
+            if ticket_type.find('Name').text.lower() == 'support':
+                cls.TICKET_TYPE_SUPPORT = ticket_type.find(
+                    'TicketTypeID').text
+                return cls.TICKET_TYPE_SUPPORT
```
I would re-factor 2 above helpers to extract common code.

RayeN commented on teamsupport/models.py:
```diff
@@ -23,23 +37,99 @@ def __init__(self, client, ticket_id=None, data=None):
+    def get_description(self):
+        ticket_actions = self.client.get_ticket_actions(
+            self.id, SystemActionTypeID=ACTION_TYPE_DESCRIPTION)
+        if len(ticket_actions) >= 1:
```
why not just `if ticket_actions:`?

Commit:
Avoid code duplication.

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Optimization

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Change attributes name.

snitch commented:
:sparkles: No lint errors found. :sparkles:

RayeN commented:
:+1: Looks good to me.

stefanor commented on integration_tests/test_tickets.py:
```diff
@@ -0,0 +1,80 @@
+    def test_new_contact_is_created(self):
+        self.assertEqual(self.contact.FirstName, self.first_name)
+        self.assertEqual(self.contact.LastName, self.last_name)
+        self.assertEqual(self.contact.Email, self.email)
```
The contact is presumably only created the first time this integration test runs.

Ideally we'd use randomised names. But then we'd need a sandbox account to run the tests in, which I assume we don't have?

stefanor replied:
Oh, we delete the contact on tearDown. Then we probably should randomise it more.

stefanor replied:
Otherwise eugene will never be able to file tickets :P

RayeN replied:
I think he'll be perfectly fine with that.

stefanor replied:
Yeah, I think this is fine. It just could be better. :+1: as-is.

ekoval replied:
> Otherwise eugene will never be able to file tickets :P

I don't have TS account at all. If Support wants to file a ticket associated with me, they will be able to create a contact with my email.

stefanor replied:
It's the email address that matters, anyway. Not the name.

ekoval replied:
Sorry, I've reviewed code again and can't see my email. I see `test@email.com`.

```
class TestTicket(unittest.TestCase):
    first_name = 'Eugene'
    last_name = 'Koval'
    email = 'test@email.com'
    ticket_name = 'test ticket dont touch'
    ticket_text = 'test descr dont touch'
```

stefanor replied:
Which is why that isn't really an issue.

stefanor commented on teamsupport/models.py:
```diff
@@ -23,23 +37,102 @@ def __init__(self, client, ticket_id=None, data=None):
+    @classmethod
+    def _get_ticket_status_new(cls):
+        if cls.TICKET_STATUS_NEW is not None:
+            return cls.TICKET_STATUS_NEW
```
Having a private method return a public constant (which isn't constant) is weird. `_ticket_status_new` may be better after all.

stefanor commented on teamsupport/models.py:
```diff
@@ -23,23 +37,102 @@ def __init__(self, client, ticket_id=None, data=None):
+
+        ticket_types = cls.get_client().get_ticket_types()
+        cls.TICKET_TYPE_SUPPORT = cls._find_element_id_by_attribute_value(
+            ticket_types, 'TicketTypeID', 'Name', 'Support')
```
`Support` also seems very specific to yola's workflow. Should this not be a parameter that gets passed in to ticket creation?

RayeN replied:
Or we should move that to configuration and read it from it

ekoval replied:
It is not specific. I tried these tests on my own account - it works.

The front page of TS says:

> Customer Service Software Built for B2B Customer **Support**.

stefanor replied:
So whenever creating a ticket, setting the type to `Support` is a reasonable thing to do? even if you aren't us?

ekoval replied:
Yes, because I tried this and it works. It was my private account. `Support` is present in ticket types list.

stefanor replied:
I wasn't asking if it works. I was asking if it's reasonable.

ekoval replied:
It makes sense to make this customizable. But not now. I'm not going to get back to TS-related code few weeks until I recover. Unless we have some critical bugs.

stefanor commented on teamsupport/models.py:
```diff
@@ -23,23 +37,102 @@ def __init__(self, client, ticket_id=None, data=None):
+    @classmethod
+    def _find_element_id_by_attribute_value(
+            cls, xml_array, id_attr, value_attr, value):
+        for xml in xml_array:
```
for element in elements?

XML is a format. Its elements aren't called XMLs

ekoval replied:
I know xml is a format. I used this name to indicate that method accepts XML.

stefanor replied:
I know you know that :) I'm just pointing out that "for xml in " reads weirdly.

ekoval replied:
Maybe `xml_elements`... As for me it's better to indicate nature of this parameter.

stefanor commented on CHANGELOG.rst:
```diff
@@ -1,6 +1,7 @@
 Changelog
 =========
 
+
```
It'd be nice to fill the changelog out as we go

RayeN replied:
I think it supposed to be populated in version bump pull.

stefanor replied:
I think we've come to realise that doing it as you go works better.

ekoval replied:
I missed that discussion. The latest news I know is a @blaix 's requirement to move version to a separate pull.

stefanor replied:
That's separate to adding changelog entries

ekoval replied:
I don't understand why at all, but ok, changed.

beck replied:
The person preparing a new release is not always the person who wrote the changes.

It is much more difficult to parse out a good summary from a diff after the fact, whereas if the original author provides their own summary, then the changelog is ready to go upon release

Commit:
Change names.

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
ChangeLog.

snitch commented:
:sparkles: No lint errors found. :sparkles:
